core/src/main/scala/kafka/log/ProducerStateManager.scala

```diff
@@ -661,8 +661,8 @@ class ProducerStateManager(val topicPartition: TopicPartition,
       None
   }
 
-  private def latestSnapshotFile: Option[File] = {
-    val files = listSnapshotFiles
+  private def latestSnapshotFile(maxSnapshotOffset: Long): Option[File] = {
+    val files = listSnapshotFiles.filter(file => offsetFromFilename(file.getName) <= maxSnapshotOffset)
     if (files.nonEmpty)
       Some(files.maxBy(file => offsetFromFilename(file.getName)))
     else
```

Comment:
You can rewrite the code in the following way:

```scala
listSnapshotFiles.map(file => (file, offsetFromFilename(file)))
  .filter { case (_, offset) => offset <= maxSnapshotOffset }
  .headOption
  .maxBy { case (_, offset) => offset }
  .map { case (file, _) => file }
```

Reply:
@ijuma Thanks for the comment. I will look into this.